### PR TITLE
Forward AD formulas batch 2

### PR DESCRIPTION
**Pull request by albanD**
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58094 Forward AD formulas batch 3
* **#57863 Forward AD formulas batch 2**

Differential Revision: [D28387763](https://our.internmc.facebook.com/intern/diff/D28387763)

### CODE REVIEW

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387763).

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -4202,6 +4205,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    dtypes=all_types_and(torch.bfloat16),
                    dtypesIfCUDA=all_types_and(torch.half, torch.bfloat16),
                    assert_autodiffed=True,
+                   supports_forward_ad=True,
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:4208]: Are all the variants of clamp being tested? I see another clamp OpInfo above that might also need to be added?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -5984,6 +5995,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     OpInfo('vstack',
            dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
            sample_inputs_func=sample_inputs_hstack_dstack_vstack,
+           supports_forward_ad=True,
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:5998]: Nice! `hstack` and `vstack` are indirectly supported because of `cat`.  You able to find out because of https://github.com/pytorch/pytorch/pull/58230/files#diff-d183f2afc51d6a59bc70094e8f476d2468c45e415500f6eb60abad955e065156R236 this line you added earlier right?

Not sure I see cross` being tested. Maybe you could modify `test_cross` in test_linalg and also method tests for cross

File: test/test_autograd.py [hunk 3]
```diff
@@ -5369,9 +5369,10 @@ def gradgradcheck_method_precision_override(test_name):
     return override
 
 def run_grad_and_gradgrad_checks(test_case, name, test_name, apply_method, output_variable,
-                                 input_variables, run_gradgradcheck=True, check_batched_grad=True):
+                                 input_variables, run_gradgradcheck=True, check_batched_grad=True,
```
Comment (soulitzer) [inline test/test_autograd.py:5372]: These changes to `test_autograd` are for method tests? Maybe you could add a similar check that checks_forward_ad=True would raise an error unless its explicitly set to True?

**Review by soulitzer — COMMENTED** [review]
Haven't looked at the actual formulas, but here are some comments on the rest

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387763).

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:4223]: If `supports_forward_ad=False` for the other variants, then we test that it raises a nice error saying this is not implemented. All the tests are green, so it should be ok.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:6040]: Yes exactly. It will actually fail if I add support for something without setting the flag to True.

Let me update the cross test!

File: tools/autograd/load_derivatives.py [hunk 4]
```diff
@@ -137,7 +137,7 @@ def postprocess_forward_derivatives(
     def find_required_inputs(formula: str, postfix: str) -> Tuple[str, ...]:
         required_inputs = set()
         for arg in args_with_derivatives:
-            if arg.type == 'TensorList':
+            if arg.type == 'at::TensorList':
```
Comment (zou3519) [inline tools/autograd/load_derivatives.py:140]: Is this a bugfix? (Should this be `arg.type == 'at::TensorList' or arg.type == 'TensorList'`?)

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -401,15 +416,18 @@
 - name: cross(Tensor self, Tensor other, int? dim=None) -> Tensor
   self: other.conj().cross(grad, dim)
   other: grad.cross(self.conj(), dim)
+  result: "at::cross(self_t, other_p, dim.has_value() ? *dim : -1) + at::cross(self_p, other_t, dim.has_value() ? *dim : -1)"
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml:419]: If dim is None... doesn't dim default "to the first dimension found with size 3?", not -1? https://pytorch.org/docs/stable/generated/torch.cross.html

File: tools/autograd/load_derivatives.py (hunk 4 above)
Comment (albanD) [reply tools/autograd/load_derivatives.py:140]: This is a bugfix indeed. This code wasn't actually used before and it is now that we have the "cat" formula.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 6]
```diff
@@ -3468,6 +3468,59 @@ Tensor lu_unpack_backward(
...
+    // The gradient we want to compute is (t0, a*t1, a*b*t1, 0, 0)
+    // We do this by:
+    // Get a mask of all zeros (0, 1, 0, 1, 0)
+    auto mask_zeros = self_p.eq(0);
+    // Get a mask of the first zero for each dim (0, 1, 0, 0, 0)
+    auto mask_first_zero = mask_zeros.logical_and(mask_zeros.cumsum(dim).eq(1));
+
+    // Get the new grad value that should be used after any zero happened:
+    // (X, a*t1, a*b*t1, 0, 0) = cumprod((a, t1, b, 0, c))
+    auto new_grad = self_p.masked_scatter(mask_first_zero, self_t.masked_select(mask_first_zero).view(-1)).cumprod(dim);
```
Comment (zou3519) [inline torch/csrc/autograd/FunctionsManual.cpp:3513]: Could this just be `torch.where(mask_first_zero == 1, self_t, self_p).cumprod(dim)` ?
mask_first_zero is (0, 1, 0, 0, 0), self_t is (t0, t1, ..., t4), self_p is (a, 0, b, 0, c).

torch.where(mask_first_zero == 1, self_t, self_p) gives us (a, t1, b, 0, c)

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -3468,6 +3468,59 @@ Tensor lu_unpack_backward(
...
+
+    // Get a mask of everything after the first zero: (0, 1, 1, 1, 1)
+    auto mask_after_first_zero = mask_first_zero.cumsum(dim);
+
+    // Do the final replacement
+    return at::where(mask_after_first_zero.to(ScalarType::Byte), new_grad, gradient);
```
Comment (zou3519) [inline torch/csrc/autograd/FunctionsManual.cpp:3519]: Does at::where not handle ScalarType::bool?

**Review by zou3519 — COMMENTED** [review]
Formula for cross looks fishy, other than that the formulas lgtm

File: tools/autograd/load_derivatives.py (hunk 4 above)
Comment (zou3519) [reply tools/autograd/load_derivatives.py:140]: Should we `assert 'TensorList' not in arg.type` after this if statement? Would make it easier to catch bugs like this in the future

File: tools/autograd/derivatives.yaml (hunk 5 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:419]: Yes I should just pass "dim" here.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 7 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:3519]: It should be a to bool indeed.

Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:3519]: Also cumsum accept a dtype argument that I might be able to use.

File: test/test_autograd.py (hunk 3 above)
Comment (albanD) [reply test/test_autograd.py:5372]: This will induce larger changes because some other tests will need to be updated as well. I'll do that in a follow up PR.

File: tools/autograd/load_derivatives.py (hunk 4 above)
Comment (albanD) [reply tools/autograd/load_derivatives.py:140]: I don't think so. It was just me not knowing that the namespace is included in this name.
I think there should be methods on the type to check that directly but I couldn't find it.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 7 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:3519]: cumsum with boolean output dtype is not implemented so I'll do the regular cumsum and then cast to bool.

**Review by zou3519 — APPROVED** [review]
Don't forget to revert the changes to the submodules, otherwise this lgtm

**Review by soulitzer — COMMENTED** [review]
Still seems weird to me that the Tensor variant of clamp does not raise an error since we added formulas for both scalar/tensor variants but only added `support_forward_ad=True` for the scalar variant in OpInfos. Any idea whats up with that?

LGTM otherwise though.

**Comment by albanD** [comment]
> Still seems weird to me that the Tensor variant of clamp does not raise an error since we added formulas for both scalar/tensor variants but only added support_forward_ad=True for the scalar variant in OpInfos. Any idea whats up with that?

We did not add the formula for the Tensor variant actually, only the multiple scalar versions. So this is expected.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387763).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387763).

**Comment by soulitzer** [comment]
> We did not add the formula for the Tensor variant actually, only the multiple scalar versions. So this is expected.

Aren't clamp_max.Tensor and clamp_min.Tensor tensor variants? Or maybe I'm missing something?

**Comment by albanD** [comment]
They are, but for the clamp_min/max functions. The Tensor variant for clamp is `clamp.Tensor` which is not getting a forward formula.
